### ISSUE TITLE
fox.com

### ISSUE THREAD

**Issue by jspenguin2017** [jspenguin2017/uBlockProtector#660]
https://www.fox.com/watch/2046d8741d77ef3febfb6004f68a55f2/

Preroll ads not blockable with just filters. Related: #87

![chrome_2017-10-08_23-10-20](https://user-images.githubusercontent.com/20338483/31319299-5c13449c-ac7e-11e7-94cd-35a1577f690b.png)

**Comment by uBlock-user** [comment]
Are preroll ads third-party ?

**Comment by jspenguin2017** [comment]
I don't know.

**Comment by uBlock-user** [comment]
Yes, they're. They're fired via an `XHR` from `uplynk.com` and nooping doesn't work on them.

**Comment by jspenguin2017** [comment]
Can you dump one of them?

**Comment by uBlock-user** [comment]
> Can you dump one of them?

I can give you the url if you mean that by dumping - 

https://stgec-ausw.uplynk.com/80C078/ausw/slices/b83/d8e6eec26bf544f084bad49a7fa2eac5/b83745da5a0746139683a3e69d663722/G0000000A.ts?pbs=a0f80247124640949fac5c3b35f4110e&si=2

https://content-ausw2.uplynk.com/vodredir4/b83745da5a0746139683a3e69d663722/G00000000.ts?ptime=31.9706666667&pbs=a0f80247124640949fac5c3b35f4110e&seq=5&o=https%3A//stgec-ausw.uplynk.com/80C078/ausw/slices/b83/d8e6eec26bf544f084bad49a7fa2eac5/b83745da5a0746139683a3e69d663722/G00000000.ts%3Fpbs%3Da0f80247124640949fac5c3b35f4110e%26si%3D2

https://content-ausw2.uplynk.com/check2?b=b83745da5a0746139683a3e69d663722&v=b83745da5a0746139683a3e69d663722&r=g&pbs=a0f80247124640949fac5c3b35f4110e

https://content.uplynk.com/preplay/910bf2f5e3394568ba2fb7cead52f5d3.json?v=2&rays=gkjihfedcb&ad=fw_prod&ad.flags=:slcb+sltp+qtcb+emcr+fbad+dtrd+exvt&ad.csid=app.dcg-foxnow%2Fweb%2Ffox&ad.kv=fnl-affiliate,,fs-affiliate,,_fw_ae,nomvpd,_fw_did_idfa,,_fw_did_google_advertising_id,,_fw_did_android_id,,_fw_did,,_fw_nielsen_app_id,,_fw_vcid2,116450:665001-1509227392-107.170.97.37,aap_truex,true,adPackage,dai&expand=&tc=1&exp=1509227632&rn=989529&ct=a&cid=910bf2f5e3394568ba2fb7cead52f5d3&sig=d90d93849d466a3620548eaf0a662f7d804b705ce767f42460d1016f3299e59b

https://content-ausc1.uplynk.com/preplay2/910bf2f5e3394568ba2fb7cead52f5d3/631fd732040ec59e6bc3fc848393be71/lzLeCtsARg4EzeXCQI61XEea2hB5h73oMpJHKsVgqF.m3u8?pbs=4870fb3964a1486e972893800794ba41


>  what do you mean by doesn't work?

Lets noop that `preplay` with `||uplynk.com/preplay^$xmlhttprequest,domain=fox.com`

Result - https://i.gyazo.com/fc9820639a359d7147492a72f42e3178.png

The website breaks ofcourse.

**Comment by uBlock-user** [comment]
Blocking `/vodredir` blocks the player from redirecting you to loading ads, unfortunately the video won't load either.

**Comment by jspenguin2017** [comment]
I mean the content of the request, does any of them return a plain text (JSON/XML/etc) format? I think the links might have expired, and my VPN is used up for this month, also I'm busy with exams, this can take a while, if you want, you can play around with `a.loopbackXHR`.

**Comment by uBlock-user** [comment]
No they get downloaded immediately and speaking of VPNs, I used Hola extension as it doesn't have any B/W caps, you can test it if you are willing to install the chrome extension.

**Comment by jspenguin2017** [comment]
OK I got it dumped: https://pastebin.com/tj08zbyK
I'll analyze it later.

**Comment by jspenguin2017** [comment]
Huh, it's kind of weird, `a.replaceXHR` can't intercept that request, and initiator is listed as "Other" in debugger. Since uBO can block it, I would be able to do a redirect in the background page, but if the request isn't static then it will be not fixable... It will be fixable in Quantum though, tricky, I need to think about it.

**Comment by jspenguin2017** [comment]
Well, it's not a total waste of time, at least I found another way that doesn't work... And fixed a performance issue in Core.

**Comment by jspenguin2017** [comment]
A quest for someone who is bored: 
Pin down the script that performs the request to `https://content.uplynk.com/preplay/(stuff).json?(stuff)`

**Comment by uBlock-user** [comment]
https://github.com/uBlockOrigin/uAssets/issues/772

Nope, nothing, nobody even bothered to take a stab at it yet..

**Comment by uBlock-user** [comment]
I was thinking of summoning okiehsch here but it looks that you're doing just fine.

**Comment by jspenguin2017** [comment]
I'm not doing just fine, I don't have a solution for Chromium.

**Comment by jspenguin2017** [comment]
Yea, the payload is dynamic, I can intercept it only in the background page, this is probably not fixable until redirect can be asynchronous.

**Comment by uBlock-user** [comment]
So what can you do at the moment ? Also is safe to say you have a workaround atleast ?

**Comment by jspenguin2017** [comment]
Nothing, I don't have a way to fix this with just extension, need a server / proxy. 
I'll make a prototype later.

**Comment by jspenguin2017** [comment]
What dark magic are they using to dodge my `replaceXHR`?

**Comment by uBlock-user** [comment]
They certainly did this keeping this extension in mind, so I guess we will have to go through each script to see what it does from the beginning.

**Comment by jspenguin2017** [comment]
It's not like you can bypass content script just by keeping an extension in mind. Although I know an exploit that bypass content script... Not sure if they are abusing that.

**Comment by jspenguin2017** [comment]
I'll close this for now, I'll see if I can make a background replace with libmicro.

**Comment by jspenguin2017** [comment]
I have a cool idea, I want to try it out.

**Comment by jspenguin2017** [comment]
![image](https://user-images.githubusercontent.com/7283682/34324056-eed5f252-e81f-11e7-95e0-c2874d89a140.png)

**Comment by BlohoJo** [comment]
>  _jspenguin2017 commented_
> OK, on Firefox only, I am able to force the player to show progress bar when playing ads.

Only for FireFox 57+ though. :\

**Comment by jspenguin2017** [comment]
@BlohoJo I get what you are saying, but I don't want to carry 20 different compatibility flags, and I have the latest Firefox installed, chances are I won't go back to an older version just to test compatibility. This is volunteer work and supporting 2 browsers is already a lot.

**Comment by BlohoJo** [comment]
Of course I understand. It's Mozilla's fault really. :P 

Hopefully the other Mozilla XUL browsers will eventually be able to update to the 57 base.

**Comment by Ravizz** [comment]
Geo-Locked for me.

**Comment by titosalah** [comment]
@jspenguin2017 @uBlock-user  Away from ads please can you help me to download english subtitle from video web page -- subtitle name like this end with extension  .SCC  (3 X 1-f4b871951085424ea77355be45c85006.en.scc) i use https://github.com/rg3/youtube-dl work with some video and not all videos work  any way to download it from developer option >> thank you very much
